mirgecom/steppers.py

```diff
@@ -136,6 +144,22 @@ def _advance_state_stepper_func(rhs, timestepper,
 
         state = timestepper(state=state, t=t, dt=dt, rhs=compiled_rhs)
 
+        if force_eval is None:
+            unevaluated_state = state
+            state = force_evaluation(actx, state)
+            if not _strict_array_equal(actx, state, unevaluated_state):
```

Comment:
I don't think it's necessary to base this on equality. You could simply traverse the array container and look for the first instance of `pytato.Array` that's not a `pytato.DataWrapper`. If you find any, it means that you would need force-eval. We could/should probably offer this predicate in `arraycontext`.